src/sql/reuse.rs

```diff
@@ -320,4 +362,173 @@ mod tests {
+            right: Box::new(Base(Literal(Literal::Integer(20.into())))),
+        });
+        let pb = ComparisonOp(ConditionTree {
+            operator: Operator::Greater,
+            left: Box::new(Base(Field(Column::from("a")))),
+            right: Box::new(Base(Literal(Literal::Integer(60.into())))),
+        });
+        let pc = ComparisonOp(ConditionTree {
+            operator: Operator::Less,
+            left: Box::new(Base(Field(Column::from("a")))),
+            right: Box::new(Base(Literal(Literal::Integer(10.into())))),
+        });
+        let pd = ComparisonOp(ConditionTree {
+            operator: Operator::Greater,
+            left: Box::new(Base(Field(Column::from("a")))),
+            right: Box::new(Base(Literal(Literal::Integer(80.into())))),
+        });
+
+        // a < 20 or a > 80
+        let cp1 = LogicalOp(ConditionTree {
+            left: Box::new(pa.clone()),
+            right: Box::new(pb.clone()),
+            operator: Operator::Or
+        });
+
+        // a < 10 or a > 60
+        let cp2 = LogicalOp(ConditionTree {
+            left: Box::new(pc),
+            right: Box::new(pd),
+            operator: Operator::Or
+        });
+
+        // a < 80 or a > 20
+        let cp3 = LogicalOp(ConditionTree {
+            left: Box::new(pb),
+            right: Box::new(pa),
+            operator: Operator::Or
+        });
+
+
+        assert!(complex_predicate_implies(&cp2, &cp1));
+        assert!(!complex_predicate_implies(&cp1, &cp2));
+        assert!(complex_predicate_implies(&cp2, &cp3));
+        assert!(!complex_predicate_implies(&cp3, &cp2));
+    }
+
+    #[test]
+    fn complex_predicate_implication_and() {
+        use nom_sql::ConditionExpression::*;
+        use nom_sql::ConditionBase::*;
+        use nom_sql::Literal;
+        let pa = ComparisonOp(ConditionTree {
+            operator: Operator::Greater,
+            left: Box::new(Base(Field(Column::from("a")))),
+            right: Box::new(Base(Literal(Literal::Integer(20.into())))),
+        });
+        let pb = ComparisonOp(ConditionTree {
+            operator: Operator::Less,
+            left: Box::new(Base(Field(Column::from("a")))),
+            right: Box::new(Base(Literal(Literal::Integer(60.into())))),
+        });
+        let pc = ComparisonOp(ConditionTree {
+            operator: Operator::Greater,
+            left: Box::new(Base(Field(Column::from("a")))),
+            right: Box::new(Base(Literal(Literal::Integer(10.into())))),
+        });
+        let pd = ComparisonOp(ConditionTree {
+            operator: Operator::Less,
+            left: Box::new(Base(Field(Column::from("a")))),
+            right: Box::new(Base(Literal(Literal::Integer(80.into())))),
+        });
+
+        // a > 20 and a < 60
+        let cp1 = LogicalOp(ConditionTree {
+            left: Box::new(pa.clone()),
+            right: Box::new(pb.clone()),
+            operator: Operator::And
+        });
```

Comment:
Sorry, my bad, too many predicates :-) `cp1` is just the flipped version of `cp3`; I misread it as being the opposite equality direction (as in `cp3` above)